modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,113 @@
+        (
+          scEvents.traverse(evt => evt.pure[F].product(StateChannelEventWithFee(evt))).map(_.toMap),
+          dagEvents.traverse(evt => evt.pure[F].product(DAGBlockEventWithFee(evt))).map(_.toMap)
+        ).flatMapN {
+          case (scEventsMap, dagEventsMap) =>
+            def removeOne(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): (List[StateChannelEvent], List[DAGEvent]) =
+              (scEvents, dagEvents) match {
+                case (Nil, Nil)    => (Nil, Nil)
+                case (_ :: t, Nil) => (t, Nil)
+                case (Nil, _ :: t) => (Nil, t)
+                case (h1 :: t1, h2 :: t2) =>
+                  Ordering[EventWithFee].compare(scEventsMap(h1), dagEventsMap(h2)) match {
+                    case n if (n == -1) || (n == 0 && t1.length >= t2.length) => (t1, dagEvents)
+                    case _                                                    => (scEvents, t2)
+                  }
+              }
+
+            val scEventsSorted = scEvents.sortWith((left, right) => Ordering[EventWithFee].lt(scEventsMap(left), scEventsMap(right)))
+            val dagEventsSorted = dagEvents.sortWith((left, right) => Ordering[EventWithFee].lt(dagEventsMap(left), dagEventsMap(right)))
+
+            (scEventsSorted, dagEventsSorted).tailRecM {
+              case (Nil, Nil) => MonadThrow[F].raiseError(new IllegalStateException("No events to cut"))
+
+              case (lefts, rights) =>
+                val result = removeOne(lefts, rights)
+                doesNotExceedMaxSize(result._1, result._2).map(flag => Either.cond(flag, result, result))
+            }
+        }
+
+      trait EventWithFee {
+        def fee: Long
+        def hash: Hash
+        def isMyChild(that: EventWithFee): Boolean
+      }
+
+      implicit val ordering: Ordering[EventWithFee] =
+        Ordering.fromLessThan[EventWithFee] { (left, right) =>
+          right.isMyChild(left) ||
+          left.fee < right.fee ||
+          (left.fee == right.fee && Ordering[Hash].lt(left.hash, right.hash))
+        }
+
+      case class StateChannelEventWithFee(event: StateChannelEvent, hash: Hash) extends EventWithFee {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+
+        def isMyChild(that: EventWithFee): Boolean =
+          that match {
+            case evt: StateChannelEventWithFee => evt.event.snapshotBinary.lastSnapshotHash === hash
+            case _                             => false
+          }
+      }
+      object StateChannelEventWithFee {
+        def apply(event: StateChannelEvent): F[StateChannelEventWithFee] =
+          event.snapshotBinary.toHashed.map(hashed => StateChannelEventWithFee(event, hashed.hash))
+      }
+
+      case class DAGBlockEventWithFee(event: Hashed[Block]) extends EventWithFee {
+        val fee: Long = event.signed.value.transactions.toIterable.map(_.value.fee.value.value).sum
+
+        val hash: Hash = event.hash
```

Comment:
Did we settle that it doesn't need to be compared by `proofsHash` (as it is compared by in other places) because this hash is only for deterministicity?

Reply:
@TheMMaciek I believe so. @ryle-goehausen, do you want to comment?